### ISSUE TITLE
Ambiguous use of `classProbabilities` in Decision Tree implementation.

### ISSUE THREAD

**Issue by RishabhGarg108** [mlpack/mlpack#2888]
## Problem
In the Decision Tree implementation, `classProbabilities` is of type `arma::vec`. It has been used to store various different kinds of informations at different places in implementation which makes it very hard to follow and it might cause bugs in future if anyone who doesn't know exactly what it does and tries to change something.

Following are the different values that `classProbabilities` during its lifetime :-
1. If the splitting dimension is a categorical attribute, then it is storing the number of categories of that dimension. This number of categories is also the number of children in this case.
2. While, if the splitting dimension is a numerical attribute, then it is storing the point about which binary split should be done. 
3. Also, the number of categories is an `int`, so while it is stored in `classProbabilities`, it is implicitly converted into a `float` and the return type of `AllCategoricalSplit::NumChildren` is `size_t`, so it then again gets implicitly converted into an `integer`. All this creates a huge mess and can cause very hard to find bugs.
4. As the name suggests, it also stores the actual probabilities for each class for leaf nodes.
5. Moreover, If in future, regression tree is implemented using this existing implementation, then `classProbabilities` will be storing the mean of all the data points of that node too.

Another similar issue is with `dimensionTypeOrMajorityClass`. Although here the name does resolve some ambiguity but still it is not good to store completely different information in same variable.

## Solution

For fixing the ambiguity of `classProbabilities`, we can create two variables -- `splitInfo` and `classProbabilities`. 
1. `splitInfo` will be a `double` which will store either the binary split point or the number of categories. Also, wherever there is an implicit casting, we will cast it explicitly with a comment as to why we are doing it. This `splitInfo` will contain information only for internal nodes of the tree.
2. `classProbabilities` will store the probabilities of each class or in future when regression trees are implemented, it will also store the mean of the data points. This information will be useful only for the leaf nodes of the tree.

Fixing `dimensionTypeOrMajorityClass` is pretty straightforward. We will split it into `dimensionType` and `majorityClass` and then use them at their appropriate places.

If anyone has some better suggestions to solve this, then please feel free to discuss here.

**Comment by rcurtin** [comment]
This is 100% intentional.  The goal is to keep `sizeof(DecisionTree<>)` small (that is, we want each decision tree node to be represented with as little memory as possible).  Importantly, that is helpful for prediction to be efficient.

Each of these members are fully commented, and that should remove the confusion:

```
  //! The vector of children.
  std::vector<DecisionTree*> children;
  //! The dimension this node splits on.
  size_t splitDimension;
  //! The type of the dimension that we have split on (if we are not a leaf).
  //! If we are a leaf, then this is the index of the majority class.
  size_t dimensionTypeOrMajorityClass;
  /**
   * This vector may hold different things.  If the node has no children, then
   * it is guaranteed to hold the probabilities of each class.  If the node has
   * children, then it may be used arbitrarily by the split type's
   * CalculateDirection() function and may not necessarily hold class
   * probabilities.
   */
  arma::vec classProbabilities;
```

Sometimes, speed comes at the cost of ease of understanding, and there is little that can be done.  This is one of those cases.

> Also, the number of categories is an int, so while it is stored in classProbabilities, it is implicitly converted into a float and the return type of AllCategoricalSplit::NumChildren is size_t, so it then again gets implicitly converted into an integer. All this creates a huge mess and can cause very hard to find bugs.

`arma::vec` holds `double`s, not `float`s, but even if it did hold `float`s, a 32-bit floating point number can represent all integers between 0 and something like 2^24 losslessly.  For doubles that exponent is I think more than twice as much.  If someone is trying to train a decision tree on data with 2^24 categories... they are going to have serious, serious problems long before `classProbabilities` even comes into the picture.  So, in the context that we can reasonably expect this to be used, it is safe.  (I understand and agree that this would not generally be considered good practice---but again, this is what we have to do to keep the size of the class small.)

**Comment by RishabhGarg108** [comment]
Hey @rcurtin , I completely agree with you. I too thought that it might have been done deliberately, but my question is whether the reduction in memory is significant or not ?

I did a bit of maths. For a full decision tree of depth 15, storing an extra double (8 bytes) will consume 256KB of extra space. Now, even if we consider an ensemble of 100 such trees, then also the total extra size would be 25MB. Now, If someone is training such a heavy ensemble with such a big tree and so many trees, then we can assume that the user will be on a high end machine with no such memory issue.

On the other hand, how does this increase in memory affects the running time that I don't know. If it does have a significant increase in time, then this justifies the complicated implementation. But if it doesn't, then I think we can simplify the implementation a bit.

Please correct me if I said something wrong.

**Comment by rcurtin** [comment]
The size matters because an L1 cache is small.

**Comment by RishabhGarg108** [comment]
Ahh..I didn't think in terms of cache memory optimization. I got your point. Thanks for the clarification. :)

Closing this one.
